File: src/services/videoStream.js
/**
 * 视频流服务 - 优化版：使用异步 I/O
 */
const fs = require('fs');
const fsPromises = require('fs').promises;
const path = require('path');
const config = require('../config');
const videoIdManager = require('../utils/videoId');
const dirPermission = require('../utils/dirPermission');

// 内容类型映射
const CONTENT_TYPE_MAP = {
  '.mp4': 'video/mp4',
  '.mkv': 'video/x-matroska',
  '.avi': 'video/x-msvideo',
  '.mov': 'video/quicktime',
  '.webm': 'video/webm'
};

// 文件状态缓存（避免频繁 stat）
const statCache = new Map();
const STAT_CACHE_TTL = 30000; // 30秒缓存

/**
 * 调试日志输出
 */
function debugLog(...args) {
  const cfg = config.getConfig();
  if (cfg.debug) {
    console.log(...args);
  }
}

/**
 * 获取视频文件路径
 * @param {string} videoId 视频ID
 * @returns {object|null} { filePath, filename, contentType } 或 null
 */
function getVideoPath(videoId) {
  const videoInfo = videoIdManager.getVideoById(videoId);
  if (!videoInfo) {
    return null;
  }
  
  const { dirName, relativePath } = videoInfo;
  const cfg = config.getConfig();
  
  // 查找目录配置
  const dirConfig = cfg.videoDirs.find(dir => dir.name === dirName);
  if (!dirConfig) {
    return null;
  }
  
  const videoDir = path.resolve(dirConfig.path);
  const filePath = path.join(videoDir, relativePath);
  
  const filename = path.basename(relativePath);
  const ext = path.extname(filename).toLowerCase();
  const contentType = CONTENT_TYPE_MAP[ext] || 'video/mp4';
  
  return { filePath, filename, contentType };
}

/**
 * 获取视频信息
 * @param {string} videoId 视频ID
 * @returns {object|null} 视频信息
 */
function getVideoInfo(videoId) {
  const videoInfo = videoIdManager.getVideoById(videoId);
  if (!videoInfo) {
    return null;
  }
  
  const pathInfo = getVideoPath(videoId);
  if (!pathInfo) {
    return null;
  }
  
  return {
    dirName: videoInfo.dirName,
    relativePath: videoInfo.relativePath,
    filename: pathInfo.filename,
    contentType: pathInfo.contentType
  };
}

/**
 * 异步获取文件状态（带缓存）
 * @param {string} filePath 文件路径
 * @returns {Promise<object|null>} 文件状态
 */
async function getFileStat(filePath) {
  const cached = statCache.get(filePath);
  if (cached && Date.now() - cached.time < STAT_CACHE_TTL) {
    return cached.stat;
  }
  
  try {
    const stat = await fsPromises.stat(filePath);
    statCache.set(filePath, { stat, time: Date.now() });
    return stat;
  } catch (err) {
    return null;
  }
}

/**
 * 同步获取文件状态（带缓存）
 * @param {string} filePath 文件路径
 * @returns {object|null} 文件状态
 */
function getFileStatSync(filePath) {
  const cached = statCache.get(filePath);
  if (cached && Date.now() - cached.time < STAT_CACHE_TTL) {
    return cached.stat;
  }
  
  try {
    const stat = fs.statSync(filePath);
    statCache.set(filePath, { stat, time: Date.now() });
    return stat;
  } catch (err) {
    return null;
  }
}

/**
 * 流式传输视频（异步版本）
 * @param {object} req Express请求对象
 * @param {object} res Express响应对象
 * @param {string} videoId 视频ID
 * @param {string} username 用户名（用于权限验证）
 */
async function streamVideoAsync(req, res, videoId, username) {
  const videoInfo = videoIdManager.getVideoById(videoId);
  if (!videoInfo) {
    return res.status(404).json({ error: '视频不存在或已过期' });
  }
  
  // 权限验证
  if (!dirPermission.hasAccess(username, videoInfo.dirName)) {
    return res.status(403).json({ error: '无权访问该视频' });
  }
  
  const pathInfo = getVideoPath(videoId);
  if (!pathInfo) {
    return res.status(404).json({ error: '视频不存在或已过期' });
  }
  
  const { filePath, contentType } = pathInfo;
  const stat = await getFileStat(filePath);
  
  if (!stat) {
    return res.status(404).json({ error: '视频文件不存在' });
  }
  
  const fileSize = stat.size;
  const range = req.headers.range;
  
  if (range) {
    // 解析 Range 请求
    const parts = range.replace(/bytes=/, '').split('-');
    const start = parseInt(parts[0], 10);
    const end = parts[1] ? parseInt(parts[1], 10) : fileSize - 1;
    const chunkSize = (end - start) + 1;
    
    // 验证范围
    if (isNaN(start) || isNaN(end) || start >= fileSize || end >= fileSize || start > end) {
      return res.status(416).json({ error: '请求范围无效' });
    }
    
    debugLog(`Stream range: ${start}-${end} (${chunkSize} bytes)`);
    
    const fileStream = fs.createReadStream(filePath, { start, end });
    
    res.writeHead(206, {
      'Content-Range': `bytes ${start}-${end}/${fileSize}`,
      'Accept-Ranges': 'bytes',
      'Content-Length': chunkSize,
      'Content-Type': contentType,
      'Cache-Control': 'public, max-age=3600',
      'Access-Control-Allow-Origin': '*'
    });
    
    fileStream.pipe(res);
    
    fileStream.on('error', (err) => {
      console.error('视频流传输失败:', err);
      if (!res.headersSent) {
        res.status(500).json({ error: '视频流传输失败' });
      }
    });
  } else {
    // 完整文件传输
    debugLog(`Stream full file: ${filePath} (${fileSize} bytes)`);
    
    const fileStream = fs.createReadStream(filePath);
    
    res.writeHead(200, {
      'Content-Length': fileSize,
      'Content-Type': contentType,
      'Accept-Ranges': 'bytes',
      'Cache-Control': 'public, max-age=3600',
      'Access-Control-Allow-Origin': '*'
    });
    
    fileStream.pipe(res);
    
    fileStream.on('error', (err) => {
      console.error('视频流传输失败:', err);
      if (!res.headersSent) {
        res.status(500).json({ error: '视频流传输失败' });
      }
    });
  }
}

/**
 * 流式传输视频（兼容同步接口）
 */
function streamVideo(req, res, videoId, username) {
  streamVideoAsync(req, res, videoId, username).catch(err => {
    console.error('视频流处理失败:', err);
    if (!res.headersSent) {
      res.status(500).json({ error: '视频流处理失败' });
    }
  });
}

/**
 * 下载视频
 * @param {object} res Express响应对象
 * @param {string} videoId 视频ID
 * @param {string} username 用户名（用于权限验证）
 */
function downloadVideo(res, videoId, username) {
  const videoInfo = videoIdManager.getVideoById(videoId);
  if (!videoInfo) {
    return res.status(404).json({ error: '视频不存在或已过期' });
  }
  
  // 权限验证
  if (!dirPermission.hasAccess(username, videoInfo.dirName)) {
    return res.status(403).json({ error: '无权访问该视频' });
  }
  
  const pathInfo = getVideoPath(videoId);
  if (!pathInfo) {
    return res.status(404).json({ error: '视频不存在或已过期' });
  }
  
  const { filePath, filename, contentType } = pathInfo;
  const stat = getFileStatSync(filePath);
  
  if (!stat) {
    return res.status(404).json({ error: '视频文件不存在' });
  }
  
  const fileSize = stat.size;
  
  debugLog(`Download: ${filename} (${fileSize} bytes)`);
  
  // 设置下载响应头
  res.writeHead(200, {
    'Content-Type': contentType,
    'Content-Length': fileSize,
    'Content-Disposition': `attachment; filename*=UTF-8''${encodeURIComponent(filename)}`,
    'Cache-Control': 'no-cache'
  });
  
  const fileStream = fs.createReadStream(filePath);
  fileStream.pipe(res);
  
  fileStream.on('error', (err) => {
    console.error('视频下载失败:', err);
    if (!res.headersSent) {
      res.status(500).json({ error: '视频下载失败' });
    }
  });
}

/**
 * 清理状态缓存
 */
function clearStatCache() {
  statCache.clear();
}

module.exports = {
  getVideoPath,
  getVideoInfo,
  streamVideo,
  downloadVideo,
  clearStatCache
};

File: go-impl/internal/service/videoStream.go
package service

import (
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	"video-server/internal/config"
	"video-server/internal/utils"
)

// 内容类型映射
var contentTypeMap = map[string]string{
	".mp4":  "video/mp4",
	".mkv":  "video/x-matroska",
	".avi":  "video/x-msvideo",
	".mov":  "video/quicktime",
	".webm": "video/webm",
}

// StatCache 文件状态缓存
type StatCache struct {
	cache map[string]*statEntry
	mu    sync.RWMutex
	ttl   time.Duration
}

type statEntry struct {
	info     os.FileInfo
	cachedAt time.Time
}

var (
	statCache     *StatCache
	statCacheOnce sync.Once
)

// GetStatCache 获取文件状态缓存
func GetStatCache() *StatCache {
	statCacheOnce.Do(func() {
		statCache = &StatCache{
			cache: make(map[string]*statEntry),
			ttl:   30 * time.Second,
		}
	})
	return statCache
}

// Get 获取文件状态
func (sc *StatCache) Get(path string) os.FileInfo {
	sc.mu.RLock()
	defer sc.mu.RUnlock()
	
	if entry, ok := sc.cache[path]; ok {
		if time.Since(entry.cachedAt) < sc.ttl {
			return entry.info
		}
	}
	return nil
}

// Set 设置文件状态
func (sc *StatCache) Set(path string, info os.FileInfo) {
	sc.mu.Lock()
	defer sc.mu.Unlock()
	
	sc.cache[path] = &statEntry{
		info:     info,
		cachedAt: time.Now(),
	}
}

// VideoStreamService 视频流服务
type VideoStreamService struct {
	statCache *StatCache
}

var (
	videoStreamService *VideoStreamService
	streamOnce         sync.Once
)

// GetVideoStreamService 获取视频流服务单例
func GetVideoStreamService() *VideoStreamService {
	streamOnce.Do(func() {
		videoStreamService = &VideoStreamService{
			statCache: GetStatCache(),
		}
	})
	return videoStreamService
}

// GetVideoPath 获取视频文件路径
func (vss *VideoStreamService) GetVideoPath(videoID string) (string, string, string) {
	videoInfo := utils.GetVideoIDManager().GetVideoByID(videoID)
	if videoInfo == nil {
		return "", "", ""
	}
	
	cfg := config.GetConfig()
	
	// 查找目录配置
	for _, dir := range cfg.VideoDirs {
		if dir.Name == videoInfo.DirName {
			videoDir, _ := filepath.Abs(dir.Path)
			filePath := filepath.Join(videoDir, videoInfo.RelativePath)
			filename := filepath.Base(videoInfo.RelativePath)
			ext := strings.ToLower(filepath.Ext(filename))
			contentType := contentTypeMap[ext]
			if contentType == "" {
				contentType = "video/mp4"
			}
			return filePath, filename, contentType
		}
	}
	
	return "", "", ""
}

// GetVideoInfo 获取视频信息
func (vss *VideoStreamService) GetVideoInfo(videoID string) map[string]interface{} {
	videoInfo := utils.GetVideoIDManager().GetVideoByID(videoID)
	if videoInfo == nil {
		return nil
	}
	
	filePath, filename, contentType := vss.GetVideoPath(videoID)
	if filePath == "" {
		return nil
	}
	
	return map[string]interface{}{
		"dirName":      videoInfo.DirName,
		"relativePath": videoInfo.RelativePath,
		"filename":     filename,
		"contentType":  contentType,
	}
}

// getFileStat 获取文件状态（带缓存）
func (vss *VideoStreamService) getFileStat(path string) (os.FileInfo, error) {
	// 先查缓存
	if info := vss.statCache.Get(path); info != nil {
		return info, nil
	}
	
	// 缓存未命中，读取文件
	info, err := os.Stat(path)
	if err != nil {
		return nil, err
	}
	
	vss.statCache.Set(path, info)
	return info, nil
}

// StreamVideo 流式传输视频
func (vss *VideoStreamService) StreamVideo(w http.ResponseWriter, r *http.Request, videoID, username string) {
	videoInfo := utils.GetVideoIDManager().GetVideoByID(videoID)
	if videoInfo == nil {
		http.Error(w, `{"error":"视频不存在或已过期"}`, http.StatusNotFound)
		return
	}
	
	// 权限验证
	if !utils.HasAccess(username, videoInfo.DirName) {
		http.Error(w, `{"error":"无权访问该视频"}`, http.StatusForbidden)
		return
	}
	
	filePath, _, contentType := vss.GetVideoPath(videoID)
	if filePath == "" {
		http.Error(w, `{"error":"视频不存在或已过期"}`, http.StatusNotFound)
		return
	}
	
	stat, err := vss.getFileStat(filePath)
	if err != nil {
		http.Error(w, `{"error":"视频文件不存在"}`, http.StatusNotFound)
		return
	}
	
	fileSize := stat.Size()
	
	// 处理 Range 请求
	rangeHeader := r.Header.Get("Range")
	if rangeHeader != "" {
		vss.streamRange(w, r, filePath, contentType, fileSize, rangeHeader)
		return
	}
	
	// 完整文件传输
	vss.streamFull(w, r, filePath, contentType, fileSize)
}

func (vss *VideoStreamService) streamRange(w http.ResponseWriter, r *http.Request, filePath, contentType string, fileSize int64, rangeHeader string) {
	// 解析 Range: bytes=start-end
	rangeStr := strings.TrimPrefix(rangeHeader, "bytes=")
	parts := strings.Split(rangeStr, "-")
	
	start, _ := strconv.ParseInt(parts[0], 10, 64)
	var end int64
	if parts[1] != "" {
		end, _ = strconv.ParseInt(parts[1], 10, 64)
	} else {
		end = fileSize - 1
	}
	
	chunkSize := end - start + 1
	
	// 验证范围
	if start >= fileSize || end >= fileSize || start > end {
		w.WriteHeader(http.StatusRequestedRangeNotSatisfiable)
		return
	}
	
	// 打开文件
	file, err := os.Open(filePath)
	if err != nil {
		http.Error(w, `{"error":"无法打开文件"}`, http.StatusInternalServerError)
		return
	}
	defer file.Close()
	
	// 定位到起始位置
	file.Seek(start, io.SeekStart)
	
	// 设置响应头
	w.Header().Set("Content-Range", fmt.Sprintf("bytes %d-%d/%d", start, end, fileSize))
	w.Header().Set("Accept-Ranges", "bytes")
	w.Header().Set("Content-Length", strconv.FormatInt(chunkSize, 10))
	w.Header().Set("Content-Type", contentType)
	w.Header().Set("Cache-Control", "public, max-age=3600")
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.WriteHeader(http.StatusPartialContent)
	
	// 传输数据
	io.CopyN(w, file, chunkSize)
}

func (vss *VideoStreamService) streamFull(w http.ResponseWriter, r *http.Request, filePath, contentType string, fileSize int64) {
	file, err := os.Open(filePath)
	if err != nil {
		http.Error(w, `{"error":"无法打开文件"}`, http.StatusInternalServerError)
		return
	}
	defer file.Close()
	
	w.Header().Set("Content-Length", strconv.FormatInt(fileSize, 10))
	w.Header().Set("Content-Type", contentType)
	w.Header().Set("Accept-Ranges", "bytes")
	w.Header().Set("Cache-Control", "public, max-age=3600")
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.WriteHeader(http.StatusOK)
	
	io.Copy(w, file)
}

// DownloadVideo 下载视频
func (vss *VideoStreamService) DownloadVideo(w http.ResponseWriter, videoID, username string) {
	videoInfo := utils.GetVideoIDManager().GetVideoByID(videoID)
	if videoInfo == nil {
		http.Error(w, `{"error":"视频不存在或已过期"}`, http.StatusNotFound)
		return
	}
	
	// 权限验证
	if !utils.HasAccess(username, videoInfo.DirName) {
		http.Error(w, `{"error":"无权访问该视频"}`, http.StatusForbidden)
		return
	}
	
	filePath, filename, contentType := vss.GetVideoPath(videoID)
	if filePath == "" {
		http.Error(w, `{"error":"视频不存在或已过期"}`, http.StatusNotFound)
		return
	}
	
	stat, err := vss.getFileStat(filePath)
	if err != nil {
		http.Error(w, `{"error":"视频文件不存在"}`, http.StatusNotFound)
		return
	}
	
	file, err := os.Open(filePath)
	if err != nil {
		http.Error(w, `{"error":"无法打开文件"}`, http.StatusInternalServerError)
		return
	}
	defer file.Close()
	
	w.Header().Set("Content-Type", contentType)
	w.Header().Set("Content-Length", strconv.FormatInt(stat.Size(), 10))
	w.Header().Set("Content-Disposition", fmt.Sprintf(`attachment; filename*=UTF-8''%s`, url.PathEscape(filename)))
	w.Header().Set("Cache-Control", "no-cache")
	w.WriteHeader(http.StatusOK)
	
	io.Copy(w, file)
}
